docs/guide/parameters.rst
Parameter overview
------------------
.. currentmodule:: time_split

Overview of parameters used by :func:`time_split.split` and :func:`plot`. Integrations such as
:func:`~integration.pandas.split_pandas` may add or remove parameters, but the base function remains the same unless
otherwise stated.

.. list-table::
   :widths: 5 9 12 75
   :header-rows: 1

   * - Name
     - Default
     - Type
     - Description

   * - ``schedule``
     - N/A
     - Valid :attr:`~types.Schedule` types:

       * :attr:`~types.DatetimeIterable`
       * pandas :ref:`offset alias <pandas:timeseries.offset_aliases>`
       * `cron <https://pypi.org/project/croniter/>`_ expression
     - Generates training dates (:attr:`.DatetimeSplitBounds.mid`). **Examples:**

       * ``['2019-05-04', '2019-05-11']`` | *Hand-picked dates.*
       * ``'7d'`` | *every 7 days, aligned to the end of the available data.*
       * ``'0 0 * * MON,FRI'`` | *every Monday and Friday at midnight.*

   * - | ``before``
       | ``after``
     - | = `'7d'`
       | = `1`
     - Valid :attr:`~types.Span` types:

       * pandas :ref:`offset alias <pandas:timeseries.offset_aliases>`
       * Literal `'all'` [#]_
       * Literal `'empty'`
       * ``int >= 1``
     - Range before/after schedule timestamps.

       The **default** ``after=1`` :ref:`stretches <sphx_glr_auto_examples_list_schedule_no_data.py>` the
       :attr:`Future data <.DatetimeSplit.future_data>` until the next `schedule` timestamp, simulating models staying
       in production until a new model takes its place. That is, ``fold[i].end = fold[i + 1].mid`` for ``after=1``.

   * - ``step``
     - = `1`
     - ``int >= 1``
     - Keep every `step:th` fold in the `schedule`. **Default (1)** =keep all.

   * - ``n_splits``
     - = `0`
     - ``int >= 0``
     - Maximum number of folds. **Default (0)** =keep all.

   * - ``available``
     - = ``None``
     - :attr:`~types.DatetimeIterable`
     - Limits ``(min, max)``, or an iterable of datetime-like types that support the built-in :py:func:`min` and
       :py:func:`max` functions. Binds `schedule` to a range.

   * - ``expand_limits``
     - = `'auto'`
     - Valid :attr:`~types.ExpandLimits` types:

       * Literal `'auto'` [#]_
       * ``bool``
       * `'round_to[<tolerance]'`
     - Expand `available` data outward to its likely `"true"` limits. Disabled if ``False``, ``True == 'auto'``.
       Use :func:`~time_split.support.expand_limits` to experiment.

       **Example**: Passing ``expand_limits='d<3h'`` expands ``(min, max)`` -limits (derived from `available`) to the nearest
       day, at most 3 hours from the original limit.

       The `tolerance` argument is optional; ``expand_limits='d'`` performs regular
       :attr:`floor(min) <pandas.Timestamp.floor>` / :attr:`ceil(max) <pandas.Timestamp.ceil>` expansion of the limits.

   * - ``filter``
     - = ``None``
     - :attr:`~types.Filter` or ``str``.
     - A callable ``(start, mid, end) -> bool``. Strings are converted using :func:`~rics.misc.get_by_full_name`.

Later folds are always [#]_ preferred. For more information about the `schedule`, `before/after` and `expand_limits`-arguments,
see the :ref:`User guide`. See the :doc:`../auto_examples/index` page for plots using the various parameter options.

.. rubric:: Footnotes

.. [#] May produce empty ranges, depending on the `schedule`. Use a :attr:`~types.Filter` to remove bad folds.
.. [#] Auto-expand_limits levels are :class:`configurable <time_split.settings.auto_expand_limits>`. Use :func:`.expand_limits` to
       experiment with `expand_limits` arguments.
.. [#] This is :attr:`configurable <time_split.settings.misc.snap_to_end>` when `schedule` and `after` are both
       :attr:`timedelta <time_split.types.TimedeltaTypes>` types.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.2.0] - 2025-12-14

### Added
- Add new `Span` literal `'empty'`; setting for `before='empty'` or `after='empty'` sets start/end to the midpoint.

## [1.1.1] - 2025-12-10

### Fixed
- Fix `snap_to_end=True` applying to CRON and explicit schedule types.

## [1.1.0] - 2025-12-10

### Added
- Implemented `snap_to_end=True` (the default) logic for offset-based `after` argument.

### Fixed
- Fix a crash in `create_explorer_link`.
- Properly align timedelta-schedules to end (instead of start) unless `snap_to_end=False`.

## [1.0.5] - 2025-12-06

### Changed
- Bump app version (docs).

## [1.0.4] - 2025-11-17

### Changed 
- Updated function `support.format_expanded_limits()`
  * Accepts arbitrary _original_ iterables
  * Returns a different message when limits are the same.
  * Input is now validated.

### Removed
- No longer adds a `NullHandler` to package root logger on import.

## [1.0.3] - 2025-11-16

### Added
- Python `3.14` is now fully tested and supported in CI/CD.

## [1.0.2] - 2025-10-25

### Fixed
- Fix crashes when using `python -OO` and some other docs issues.

## [1.0.1] - 2025-08-22

### Changed
- Bump app version.

## [1.0.0] - 2025-08-21

### Added
- New sequence-like type `LogSplitProgress`; returned by `log_split_progress` instead of generic iterable.

### Changed
- Convert `settings.filter` to a keyword argument.

## [0.7.1] - 2025-07-17

### Added
- app: Add `time_split.app.reexport`.

## [0.7.0] - 2025-07-16

### Added
- The `time-split[app]` extra; adds `time_split_app` namespace.
  * New `time-split` cli; `python -m time_split`.
- The `create_explorer_link()` function will now hex-encode `bytes` data.
- Add `settings.misc.round_limits=False`. Enable to allow inward rounding of limits.

### Changed
- Make `types.SplitProgressExtras` generic on `MetricsType`.
- Allow passing `Series/DataFrame` as `plot(available=...)`.
- The `to_string()` function now supports using _.delta_ and _.iso_ (like _.auto_) in the format.

## [0.6.0] - 2024-08-31

### Changed
- Move the [Streamlit application](https://time-split.streamlit.app/) into a 
  separate repo: https://github.com/rsundqvist/time-fold-explorer/.

### Removed
- The `time_split.streamlit` submodule.
- Deprecated alias `flex`; use `expand_limits` instead.

## [0.5.0] - 2024-07-06

### Added
- Add `settings.misc.filter` and `ignore_filters` argument.
- A new Streamlit application (see `time_split.streamlit`): https://hub.docker.com/r/rsundqvist/time-split/.

### Changed
- The `plot()`-function now treats data of length two as a limits-tuple for display purposes.
- Reduce the number of the INFO-level limits expansion messages emitted.

## [0.4.0] - 2024-06-17

### Deprecated
- The `flex` argument is now called `expand_limits` in all user-facing functions. The support function with the same 
  name now uses `spec` for the equivalent argument.

### Fixed
- Remove uses of deprecated functions.

## [0.3.0] - 2024-05-18

### Added
- The `log_split_progress()`-function now accepts an optional _get_metrics()_ callback method. Metrics returned are 
  formatted based on type (if using the new `default_metrics_formatter()`), then added to the fold-end log message.
- New property `settings.log_split_progress.FORMAT_METRICS`; formats _get_metrics()_ output.
- Added several new `time_split.types` members, all related to `log_split_progress()`.

### Changed
- Module `integration.split_data` is now `integration.base` to better reflect intent. Member names are unchanged.

### Removed
- The `settings.log_split_progress.SECONDS_FORMATTER` setting no longer accepts `str` arguments.
- The `log_split_progress(extra)`-argument is no longer mutable.

## [0.2.1] - 2024-04-20

### Fixed
- Various documentation issues; broken links, naming/formatting conformity.

## [0.2.0] - 2024-04-20

### Added
* [Parameter overview](https://time-split.readthedocs.io/en/latest/guide/parameters.html) table to docs.

### Changed
* Link to relevant examples on the
  [Before and after arguments](https://time-split.readthedocs.io/en/latest/guide/spans.html)-page.
* Expose more internals in the `time_split.support` module (unstable API).
* Rename parts of the internal (unstable) API.

### Fixed
* Fix project description.
* Various other documentation issues.

## [0.1.0] - 2024-04-14

* Branch from [rics@v4.0.1](https://github.com/rsundqvist/rics/blob/v4.0.1/CHANGELOG.md).

### Changed
* Move out of `rics` namespace.

### Fixed
* Fixed a few documentation and examples issues.

[Unreleased]: https://github.com/rsundqvist/time-split/compare/v1.2.0...HEAD
[1.2.0]: https://github.com/rsundqvist/time-split/compare/v1.1.1...v1.2.0
[1.1.1]: https://github.com/rsundqvist/time-split/compare/v1.1.0...v1.1.1
[1.1.0]: https://github.com/rsundqvist/time-split/compare/v1.0.5...v1.1.0
[1.0.5]: https://github.com/rsundqvist/time-split/compare/v1.0.4...v1.0.5
[1.0.4]: https://github.com/rsundqvist/time-split/compare/v1.0.3...v1.0.4
[1.0.3]: https://github.com/rsundqvist/time-split/compare/v1.0.2...v1.0.3
[1.0.2]: https://github.com/rsundqvist/time-split/compare/v1.0.1...v1.0.2
[1.0.1]: https://github.com/rsundqvist/time-split/compare/v1.0.0...v1.0.1
[1.0.0]: https://github.com/rsundqvist/time-split/compare/v0.7.1...v1.0.0
[0.7.1]: https://github.com/rsundqvist/time-split/compare/v0.7.0...v0.7.1
[0.7.0]: https://github.com/rsundqvist/time-split/compare/v0.6.0...v0.7.0
[0.6.0]: https://github.com/rsundqvist/time-split/compare/v0.5.0...v0.6.0
[0.5.0]: https://github.com/rsundqvist/time-split/compare/v0.4.0...v0.5.0
[0.4.0]: https://github.com/rsundqvist/time-split/compare/v0.3.0...v0.4.0
[0.3.0]: https://github.com/rsundqvist/time-split/compare/v0.2.1...v0.3.0

src/time_split/settings.py
"""Global settings for the splitting logic."""

import typing as _t

import pandas as _pd

from . import types as _tst


class auto_expand_limits:  # noqa: N801
    """Configuration for the `'auto'` :attr:`~time_split.types.ExpandLimits` logic.

    This class determines how ``(min, max)``-tuples are expanded when expanded limits are enabled.
    """

    SANITY_CHECK: bool = True
    """If ``True``, use original limits if expanded limits do not pass sanity checks."""

    @classmethod
    def set_level(
        cls,
        level: _t.Literal["hour", "day"],
        *,
        start_at: _tst.TimedeltaTypes,
        round_to: _tst.TimedeltaTypes,
        tolerance: _tst.TimedeltaTypes,
    ) -> None:
        """Set a :class:`Level` used by the auto-expand_limits logic.

        Inputs are not verified until an actual split is made.

        Args:
            level: The level to set; `'hour'` or `'day'`.
            start_at: Span size at which this level starts.
            round_to: Frequency to round the range limits to.
            tolerance: Maximum amount by which to alter limits.

        Raises:
            AttributeError: For unknown `level` names.

        """
        if level not in {"hour", "day"}:
            raise AttributeError(f"Bad {level=}: Expected one of ('hour', 'day').")
        level_tuple = (start_at, round_to, tolerance)
        setattr(cls, level, level_tuple)

    class Level(_t.NamedTuple):
        """Level type used by :attr:`auto_expand_limits`."""

        start_at: _tst.TimedeltaTypes
        """Span size at which this level starts."""

        round_to: _tst.TimedeltaTypes
        """Frequency to round the range limits to."""

        tolerance: _tst.TimedeltaTypes
        """Maximum amount by which to alter limits."""

    hour: Level = Level("6 hours", "hour", "15 min")
    """Conditions under which bounds are rounded to the nearest hour.

    Default setting:
        Round to hour if the total range is at least 6 hours, but do not move the bounds more than 15 minutes.
    """

    day: Level = Level("2 days", "day", "3 hours")
    """Conditions under which bounds are rounded to the nearest day.

    Default setting:
        Round to day if the total range is at least 2 days, but do not move the bounds more than 3 hours.
    """


class plot:  # noqa: N801
    """Global settings for the :func:`.time_split.plot`-function."""

    THOUSANDS_SEPARATOR: str = "'"
    """Separator to use when printing `bar_labels`."""
    THOUSANDS_SEPARATOR_CUTOFF: int = 10_000
    """Minimum value before `bar_labels` include a :attr:`THOUSANDS_SEPARATOR`."""
    ROW_UNIT: str = "rows"
    """Unit to append to the count when displaying number of rows on the bars."""

    DATA_LABEL: str = "Data"
    """Label of the blue bar; data in the the [:attr:`~.DatetimeSplitBounds.start`, :attr:`~.types.DatetimeSplitBounds.mid`)-range."""
    FUTURE_DATA_LABEL: str = "Future data"
    """Label of the red bar; data in the the [:attr:`~.DatetimeSplitBounds.mid`, :attr:`~.types.DatetimeSplitBounds.end`)-range."""

    DEFAULT_TIME_UNIT: str = "hours"
    """Time unit to use by default when ``bar_labels=True`` and ``available=None`` or a limits tuple."""

    REMOVED_FOLD_STYLE: _t.ClassVar[dict[str, _t.Any]] = {
        "alpha": 0.35,
        "height": 0.6,
    }
    """Keyword arguments used to distinguish filtered folds when plotting with ``show_removed=True``.

    See :func:`matplotlib.pyplot.bar` for details.
    """


class log_split_progress:  # noqa: N801
    """Global settings for the :func:`.log_split_progress`-function."""

    FOLD_FORMAT: str = "'{start.auto}' <= [schedule: '{mid.auto}' ({mid:%A})] < '{end.auto}'"
    """Pretty-printed `fold`-key for other messages.

    * Only the ``start``, ``mid``, and ``end`` keys are available (see
      :attr:`~time_split.types.DatetimeSplitBounds`). You may use ``<key>.auto`` to format as a date when
      time is zero (this is the default).

    .. code-block:: python
       :caption: Sample output.

       '2021-12-30' <= [schedule: '2022-01-04' (Tuesday)] < '2022-01-04 18:00:00'
    """

    SECONDS_FORMATTER: _t.Callable[[float], str] | None = None
    """A callable ``(seconds) -> formatted_seconds``. Use :func:`rics.strings.format_seconds` if ``None``.

    Both ``seconds`` and ``formatted_seconds`` will be available to the :attr:`END_MESSAGE` message.
    """

    START_MESSAGE: str = "Begin fold {n}/{n_splits}: {fold}."
    """Message indicating that the current fold has been yielded to the user.

    Has access to all keys from the previous section, as well as:

    * The ``fold`` key (see :attr:`FOLD_FORMAT`), and
    * The ``n`` key, which is the 1-based position of the fold in `splits`, and
    * The ``n_folds`` key, which is just ``len(splits)``.

    .. code-block:: python
       :caption: Sample output.

        Begin fold 5/7: '2021-12-30' <= [schedule: '2022-01-04' (Tuesday)] < '2022-01-04 18:00:00'.
    """

    END_MESSAGE: str = "Finished fold {n}/{n_splits}: [schedule: '{mid.auto}' ({mid:%A})] after {formatted_seconds}."
    """Message indicating that the user is done with the current fold.

    Has access to all keys from the previous sections, as well as:

    * The ``seconds`` key, which is the (fractional) time the user spent in the fold, and
    * The ``formatted_seconds`` key, obtained using the :attr:`SECONDS_FORMATTER`.

    The value of ``seconds`` is obtained using :py:func:`time.perf_counter`.

    .. code-block:: python
       :caption: Sample output.

        Finished fold 5/7: [schedule: '2022-01-04' (Tuesday)] after 5m 21s.
    """

    AUTO_DATE_FORMAT = "%Y-%m-%d"
    """Short-form timestamp format_spec used by ``<key>.auto``."""
    AUTO_DATETIME_FORMAT = "%Y-%m-%d %H:%M:%S"
    """Long-form timestamp format_spec used by ``<key>.auto``."""

    FORMAT_METRICS: _tst.FormatMetrics[_t.Any] | None = None
    """A callable ``(end_message, metrics) -> str``. Use the :func:`default formatter <.support.default_metrics_formatter>` if ``None``."""


class misc:  # noqa: N801
    """Miscellaneous settings that don't fit in any other namespace."""

    snap_to_end: bool = True
    """Controls alignment of timedelta-based splits.

    .. note::

       Active only when both `schedule` and `after` are :attr:`~time_split.types.TimedeltaTypes`.

    If ``True``, ensure that the `Future data` of the final fold ends at the rightmost edge of the available data range.
    Otherwise, alignment is determined by :func:`pandas.date_range`.
    """

    filter: _t.Callable[[_pd.Timestamp, _pd.Timestamp, _pd.Timestamp], bool] | None = None
    """A callable ``(start, mid, end) -> bool`` applied to all generated folds."""
